specs/CustomDownload.md

```diff
@@ -0,0 +1,771 @@
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_SERVER_NO_RANGE,
+  /// Server does not have the requested data.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_SERVER_BAD_CONTENT,
+  /// Server did not authorize access to resource.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_SERVER_UNAUTHORIZED,
+  /// Server certificate problem.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_SERVER_CERT_PROBLEM,
+  /// Server access forbidden.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_SERVER_FORBIDDEN,
+  /// Unexpected server response. Responding server may not be intended server.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_SERVER_UNREACHABLE,
+  /// Server sent fewer bytes than the content-length header. Content-length
+  /// header may be invalid or connection may have closed. Download is treated
+  /// as complete unless there are strong validators present to interrupt the
+  /// download.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_SERVER_CONTENT_LENGTH_MISMATCH,
+  /// Unexpected cross-origin redirect.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_SERVER_CROSS_ORIGIN_REDIRECT,
+
+  /// User canceled the download.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_USER_CANCELED,
+  /// User shut down the WebView.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_USER_SHUTDOWN,
+  /// User paused the download.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_USER_PAUSED,
+
+  /// WebView crashed.
+  COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON_DOWNLOAD_PROCESS_CRASHED,
+} COREWEBVIEW2_DOWNLOAD_INTERRUPT_REASON;
+
+[uuid(9aab8652-d89f-408d-8b2c-1ade3ab51d6d), object, pointer_default(unique)]
+interface ICoreWebView2Settings2 : ICoreWebView2Settings {
+
+  /// The host may set this flag to hide the default download dialog. The CoreWebView2.DownloadStarting event will continue to fire, and the download will progress as normal, there will just be no default download UI.
+  [propget] HRESULT ShouldDisplayDefaultDownloadDialog(
+    [out, retval] BOOL* shouldDisplayDefaultDownloadDialog);
+
+  /// Sets the `ShouldDisplayDefaultDownloadDialog` property.
+  [propput] HRESULT ShouldDisplayDefaultDownloadDialog(
+    [in] BOOL shouldDisplayDefaultDownloadDialog);
+}
+
+[uuid(E8495255-938A-4784-9EAC-A4BEC8869872), object, pointer_default(unique)]
+interface ICoreWebView2_3 : ICoreWebView2_2 {
+
+  /// Add an event handler for the `DownloadStarting` event. This event is
+  /// raised when a download has begun, blocking the default download dialog,
+  /// but not blocking the progress of the download.
+  ///
+  /// The host can choose to cancel a download and change the result file path.
+  /// If the host chooses to cancel the download, the download is not saved and
+  /// no dialog is shown. Otherwise, the download is saved after the event completes,
+  /// and default download dialog is shown if the host did not choose to hide it.
+  /// The host can change the visibility of the download dialog using the
+  /// `ShouldDisplayDefaultDownloadDialog` property on `ICoreWebView2Settings`.
```

Comment:
The setting is per page navigation - ensure this is noted in the documentation for the setting. The handled property on the event args could be used to make this toggleable per download.

Reply:
Remove setting in favor of the handled property in order to avoid the case of setting the property and not adding an event handler for downloadstarting and accidentally still silently allowing all downloads.